### PR TITLE
Export grass CRS definitions as WKT2 strings instead of proj strings wherever possible

### LINKED ISSUE #18596: GRASS/Processing outputs CRS mismatch
Author Name: **Paolo Cavallini** (@pcav)
Original Redmine Issue: [10137](https://issues.qgis.org/issues/10137)
Affected QGIS version: 3.4.1
Redmine category:processing/grass

---
Apparently Processing, during the analyses starting from layers in a CRS with +towgs parameters (e.g. 23030) creates outputs with a similar projection, but without the parameters; a custom CRS is thus created, and the output does not align with the input.
This happens at least with GRASS and R backends.

---

Related issue(s): #20098 (relates), #25391 (duplicates)
Redmine related issue(s): [11884](https://issues.qgis.org/issues/11884), [17494](https://issues.qgis.org/issues/17494)

---

### PR DESCRIPTION
**Pull request by nyalldawson**
Because proj strings are lossy

Fixes #18596

### CODE REVIEW

**Comment by nyalldawson** [comment]
@gioman can you give this one a test?

**Comment by nyalldawson** [comment]
This should work for everything we can fix. There's some algorithms we can't use wkt strings for (E.g.  https://grass.osgeo.org/grass74/manuals/r.tileset.html, https://grass.osgeo.org/grass78/manuals/v.in.wfs.html), as those ONLY accept proj4 definitions for a crs style parameter. This needs to be fixed in grass before we can handle in qgis, and we'd need to do it conditionally based on whatever version the fix lands in GRASS.

**Comment by gioman** [comment]
> @gioman can you give this one a test?

 @nyalldawson will do asap!

File: python/plugins/processing/algs/grass7/Grass7Algorithm.py [hunk 1]
```diff
@@ -1031,23 +1033,25 @@ def setSessionProjectionFromProject(self):
         We creates a PROJ4 definition which is transmitted to Grass
         """
         if not Grass7Utils.projectionSet and iface:
-            self.destination_crs = iface.mapCanvas().mapSettings().destinationCrs()
-            proj4 = iface.mapCanvas().mapSettings().destinationCrs().toProj()
-            command = 'g.proj -c proj4="{}"'.format(proj4)
-            self.commands.append(command)
-            Grass7Utils.projectionSet = True
+            self.setSessionProjection(iface.mapCanvas().mapSettings().destinationCrs())
```
Comment (rouault) [inline python/plugins/processing/algs/grass7/Grass7Algorithm.py:1036]: the comment 2 line about about PROJ4 definition should be changed

**Comment by gioman** [comment]
> @gioman can you give this one a test?

@nyalldawson working fine here on master!

**Comment by nyalldawson** [comment]
@gioman do you really mean master? Or did you test this branch?

**Comment by gioman** [comment]
> @gioman do you really mean master? Or did you test this branch?

@nyalldawson I had no mean to compile QGIS today, so I quickly tested applying the patch on QGIS master on Windows (just to be sure to test the most widely used platform). I can compile and test on Linux the branch tomorrow.

**Comment by nyalldawson** [comment]
Ah, sorry I misunderstood you to mean that things were fixed already on master without this patch!

So you're +1 to merge then?

**Comment by gioman** [comment]
> So you're +1 to merge then? 

@nyalldawson yes!

**Comment by neteler** [comment]
Please note the new PR https://github.com/OSGeo/grass/pull/1240

**Comment by nyalldawson** [comment]
Thanks @neteler !

As far as I can see the other missing part is support for WKT crs definitions in the r.tileset and v.in.wfs tools. (I was planning on submitting a GRASS bug report for those after getting this PR finished)

**Comment by nyalldawson** [comment]
Cherry-picked to https://github.com/qgis/QGIS/pull/40712

**Pull request merged**